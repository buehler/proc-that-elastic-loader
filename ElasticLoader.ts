import {ILoad} from 'proc-that/dist/interfaces/ILoad';

let Promise = require('es6-promise').Promise;
let elasticsearch = require('elasticsearch');

export class ElasticLoader implements ILoad {
    private esClient:any;

    constructor(config:any, private index:string, private type:string, private idSelector:(obj:any) => any = o => o.id, private predicate:(obj:any) => boolean = o => true) {
        this.esClient = new elasticsearch.client(config);
    }

    write(object:any):Promise<void> {
        if (!this.predicate(object)) {
            return Promise.resolve();
        }

        return this.esClient.index({
            index: this.index,
            type: this.type,
            id: this.idSelector(object),
            body: object
        });
    }
}