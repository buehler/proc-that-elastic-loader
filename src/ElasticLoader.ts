import { Loader } from 'proc-that';
import { Observable } from 'rxjs';

import { Buffer } from './helpers/Buffer';

const elasticsearch = require('elasticsearch');

class NoIdProvidedError extends Error {
    constructor(public object: any) {
        super('No id provided by object');
    }
}

export class ElasticLoader implements Loader {
    private esClient: any;
    private buffer: Buffer<any> = new Buffer();

    constructor(
        config: any,
        private index: string,
        private type: string,
        private predicate: (obj: any) => boolean = () => true,
        private idSelector: (obj: any) => any = o => o.id,
    ) {
        const esConfig = JSON.parse(JSON.stringify(config));
        if (!esConfig.requestTimeout) {
            // set requestTimeout to 5min.
            // reason: when you shoot many index requests to the esClient, elasticsearch buffers your requests.
            // after the default timeout of 30s is exceeded, you receive a TimeoutError and the whole index process fails.
            esConfig.requestTimeout = 1000 * 60 * 5;
        }
        this.esClient = new elasticsearch.Client(esConfig);
        if (esConfig.maxSockets) {
            this.buffer = new Buffer(esConfig.maxSockets);
        }
    }

    public write(object: any): Observable<any> {
        if (!this.predicate(object)) {
            return Observable.empty();
        }

        const id = this.idSelector(object);

        if (id === null || id === undefined) {
            return Observable.throw(new NoIdProvidedError(object));
        }

        const promise = this.buffer
            .read()
            .then(() => this.esClient.index({
                id,
                index: this.index,
                type: this.type,
                body: object,
            }));

        this.buffer.write(object);

        return Observable.fromPromise(promise);
    }
}
