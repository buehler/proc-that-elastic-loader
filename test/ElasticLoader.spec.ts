import { Observable } from 'rxjs/Observable';

import { ElasticLoader } from '../src/ElasticLoader';

describe('ElasticLoader', () => {

    let loader: ElasticLoader;
    let client: any;

    beforeEach(() => {
        client = {
            index: o => Observable.of(o)
        };

        client.index = jest.fn(client.index);
    });

    it('should resolve on correct usage', done => {
        loader = new ElasticLoader({}, 'testIndex', 'testType');
        (loader as any).esClient = client;

        loader.write({id: 1, text: 'test'}).subscribe(null, done, done);
    });

    it('should use correct index', done => {
        loader = new ElasticLoader({}, 'testIndex', 'testType');
        (loader as any).esClient = client;

        loader.write({id: 1, text: 'test'}).subscribe(null, done, () => {
            try {
                expect(client.index.mock.calls.length).toBe(1);
                expect(client.index.mock.calls[0][0]).toMatchSnapshot();
                done();
            } catch (e) {
                done(e);
            }
        });
    });

    it('should use correct type', done => {
        loader = new ElasticLoader({}, 'testIndex', 'testType');
        (loader as any).esClient = client;

        loader.write({id: 1, text: 'test'}).subscribe(null, done, () => {
            try {
                expect(client.index.mock.calls.length).toBe(1);
                expect(client.index.mock.calls[0][0]).toMatchSnapshot();
                done();
            } catch (e) {
                done(e);
            }
        });
    });

    it('should use correct id', done => {
        loader = new ElasticLoader({}, 'testIndex', 'testType');
        (loader as any).esClient = client;

        loader.write({id: 1, text: 'test'}).subscribe(null, done, () => {
            try {
                expect(client.index.mock.calls.length).toBe(1);
                expect(client.index.mock.calls[0][0]).toMatchSnapshot();
                done();
            } catch (e) {
                done(e);
            }
        });
    });

    it('should use correct body', done => {
        loader = new ElasticLoader({}, 'testIndex', 'testType');
        (loader as any).esClient = client;

        loader.write({id: 1, text: 'test'}).subscribe(null, done, () => {
            try {
                expect(client.index.mock.calls.length).toBe(1);
                expect(client.index.mock.calls[0][0]).toMatchSnapshot();
                done();
            } catch (e) {
                done(e);
            }
        });
    });

    it('should be called on truthly predicate', done => {
        loader = new ElasticLoader({}, 'testIndex', 'testType', o => o.text === 'test');
        (loader as any).esClient = client;

        loader.write({id: 1, text: 'test'}).subscribe(null, done, () => {
            try {
                expect(client.index.mock.calls.length).toBe(1);
                done();
            } catch (e) {
                done(e);
            }
        });
    });

    it('should not called on falsly predicate', done => {
        loader = new ElasticLoader({}, 'testIndex', 'testType', o => o.text !== 'test');
        (loader as any).esClient = client;

        loader.write({id: 1, text: 'test'}).subscribe(null, done, () => {
            try {
                expect(client.index.mock.calls.length).toBe(0);
                done();
            } catch (e) {
                done(e);
            }
        });
    });

    it('should be called with correct id when using custom selector', done => {
        loader = new ElasticLoader({}, 'testIndex', 'testType', o => true, o => o.myId);
        (loader as any).esClient = client;

        loader.write({myId: 1, text: 'test'}).subscribe(null, done, () => {
            try {
                expect(client.index.mock.calls.length).toBe(1);
                expect(client.index.mock.calls[0][0]).toMatchSnapshot();
                done();
            } catch (e) {
                done(e);
            }
        });
    });

    it('should reject when no id is provided', done => {
        loader = new ElasticLoader({}, 'testIndex', 'testType');
        (loader as any).esClient = client;

        loader.write({myId: 1, text: 'test'}).subscribe(null, () => {
            done();
        }, () => {
            done(new Error('did not throw.'));
        })
    });

})
;
