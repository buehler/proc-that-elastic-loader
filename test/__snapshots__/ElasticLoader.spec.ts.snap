// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElasticLoader should be called with correct id when using custom selector 1`] = `
Object {
  "body": Object {
    "myId": 1,
    "text": "test",
  },
  "id": 1,
  "index": "testIndex",
  "type": "testType",
}
`;

exports[`ElasticLoader should use correct body 1`] = `
Object {
  "body": Object {
    "id": 1,
    "text": "test",
  },
  "id": 1,
  "index": "testIndex",
  "type": "testType",
}
`;

exports[`ElasticLoader should use correct id 1`] = `
Object {
  "body": Object {
    "id": 1,
    "text": "test",
  },
  "id": 1,
  "index": "testIndex",
  "type": "testType",
}
`;

exports[`ElasticLoader should use correct index 1`] = `
Object {
  "body": Object {
    "id": 1,
    "text": "test",
  },
  "id": 1,
  "index": "testIndex",
  "type": "testType",
}
`;

exports[`ElasticLoader should use correct type 1`] = `
Object {
  "body": Object {
    "id": 1,
    "text": "test",
  },
  "id": 1,
  "index": "testIndex",
  "type": "testType",
}
`;
